README.md
# simple_bloom_filter 🎯

A simple Dart implementation of a bloom filter. [Bloom filters](https://en.wikipedia.org/wiki/Bloom_filter) are a space-efficient probabilistic data structure that is used to test whether an element is a member of a set.

## usage

```dart
final bloom = SimpleBloomFilter(10000);
bloom.add('hello world');
if (bloom.check('hello world') == true) {
    print('"hello world" PROBABLY exists...');
}
```

## comparison to contains (see test)
...566 (0): Setting up the test...
...576 (10): Starting Bloom search 5000 elements in list with 50000 elements
...579 (3): found 2780 possible matches, 2220 are not needed to check
...579 (0): Starting normal search 5000 elements in list with 50000 elements
...831 (252): found 347 matches
✓ A group of tests Speed check

Using a simple bloom configuration where the bit == the test list size
5k searches in 50k items takes 252ms, using bloom it can say that half (2220) are not needed to check in 3ms!. So it can save you 123ms in this example. 
When the size of the to-search list increases, bloom becomes more and more usefull!

## Features

- Choose size of bit array

## Features to come

- Better hash functions
- Choose false positive probability

CHANGELOG.md
## 1.0.0

- Nullsafety
- Renamed to SimpleBloomFilter
- Added addAll
- Use Objects instead of Strings
- Removed hashList and use hashcode instead to improve performance drastically

## 0.1.3

- Fix name in README

## 0.1.2

- Update README
- Generate dartdoc

## 0.1.1

- Update description

## 0.1.0

- Hashes are random numbers
- Input must be String type


test/bloom_filter_test.dart
import 'package:bloom_filter/bloom_filter.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    BloomFilter bloom;

    setUp(() {
      bloom = BloomFilter(10000, 3);
    });

    test('First Test', () {
      bloom.add('hello world');
      bloom.add('this is just a test');
      bloom.add('Dart is fun!');
      bloom.add(1234567890.toString());
      bloom.add('{"first-name": "Dart", "last-name": "lang"}');

      expect(bloom.check('hello world'), isTrue);
      expect(bloom.check('this is just a test'), isTrue);
      expect(bloom.check('Dart is fun!'), isTrue);
      expect(bloom.check(1234567890.toString()), isTrue);
      expect(bloom.check('{"first-name": "Dart", "last-name": "lang"}'), isTrue);
      expect(bloom.check('{"first-name": "Go", "last-name": "lang"}'), isFalse);
    });
  });
}

test/simple_bloom_filter_test.dart
import 'dart:math';

import 'package:simple_bloom_filter/simple_bloom_filter.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    test('First Test', () {
      final bloom = SimpleBloomFilter(10000);
      bloom.add('hello world');
      bloom.add('this is just a test');
      bloom.add('Dart is fun!');
      bloom.add(1234567890);
      bloom.add('{"first-name": "Dart", "last-name": "lang"}');

      expect(bloom.check('hello world'), isTrue);
      expect(bloom.check('this is just a test'), isTrue);
      expect(bloom.check('Dart is fun!'), isTrue);
      expect(bloom.check(1234567890), isTrue);
      expect(bloom.check('{"first-name": "Dart", "last-name": "lang"}'), isTrue);
      expect(bloom.check('{"first-name": "Go", "last-name": "lang"}'), isFalse);
    });

    int? _lastTime;
    void _printTime(String message) {
      final now = DateTime.now().millisecondsSinceEpoch;
      print('$now (${now - (_lastTime ?? now)}): $message');
      _lastTime = now;
    }

    // ...566 (0): Setting up the test...
    // ...576 (10): Starting Bloom search 5000 elements in list with 50000 elements
    // ...579 (3): found 2780 possible matches, 2220 are not needed to check
    // ...579 (0): Starting normal search 5000 elements in list with 50000 elements
    // ...831 (252): found 347 matches
    // ✓ A group of tests Speed check
    test('Speed check', () {
      _printTime('Setting up the test...');
      final random = Random();
      const testSize = 50000;
      const matchRate = 0.1;
      final testList = List.generate(testSize, (index) => testSize + index * random.nextInt(4))..shuffle(random);
      const searchSize = 5000;
      final searhList = List.generate(searchSize, (index) => random.nextInt((testSize / matchRate).floor()));
      final bloom = SimpleBloomFilter(testSize);
      _printTime('Starting Bloom search $searchSize elements in list with $testSize elements');
      bloom.addAll(testList);
      final toCheck = searhList.where(bloom.check).length;
      _printTime('found $toCheck possible matches, ${searchSize - toCheck} are not needed to check');
      _printTime('Starting normal search $searchSize elements in list with $testSize elements');
      final total = searhList.where((e) => testList.contains(e)).length;
      _printTime('found $total matches');
    });
  });
}


lib/src/bloom_filter_base.dart
import 'dart:convert';
import 'dart:math';

class BloomFilter {
  final int _size;
  final int _hashNum;
  List<int> hashList;
  List<bool> bitArray;
  final int _min;
  final int _max;
  //m number of bits in the array
  //n the elements
  //k number of hash functions
  BloomFilter(size, hashCount)
    : _size = size,
     bitArray = List.filled(size, false),
     _hashNum = hashCount,
     _min = 1000,
     _max = 9999;

  void generateHashes() {
    hashList = List(_hashNum);
    for(var i = 0; i < _hashNum; i++) {
      hashList[i] = _min + Random().nextInt(_max - _min);
    }
  }

  BigInt _hash(String key, int hashInt) {
    var hash = BigInt.from(hashInt);
    var keyList = utf8.encode(key);
    int char;
    for (char in keyList) {
      hash = ((hash << 5) + hash) + BigInt.from(char);
    }
    return hash;
  }

  void add(String key) {
    if (hashList == null) {
      generateHashes();
    }
    for (var hash in hashList) {
      bitArray[(_hash(key, hash) % BigInt.from(_size)).toInt()] = true;
    }
  }

  bool check(String key) {
    if (hashList == null) {
      return false;
    }
    for (var hash in hashList) {
      if (bitArray[(_hash(key, hash) % BigInt.from(_size)).toInt()]) {
        return true;
      }
    }
    return false;
  }
}

example/simple_bloom_filter_example.dart
import 'package:simple_bloom_filter/simple_bloom_filter.dart';

void main() {
  final bloom = SimpleBloomFilter(10000);

  bloom.add('hello world');
  bloom.add(1234567890);
  bloom.add('{"first-name": "Dart", "last-name": "lang"}');

  if (bloom.check('hello world') == true) {
    print('"hello world" PROBABLY exists...');
  } else {
    print('"hello world" DEFINITELY does not exist...');
  }
  if (bloom.check(1234567890) == true) {
    print('"1234567890" PROBABLY exists...');
  } else {
    print('"1234567890" DEFINITELY does not exist...');
  }
  if (bloom.check('{"first-name": "Dart", "last-name": "lang"}') == true) {
    print('"{"first-name": "Dart", "last-name": "lang"}" PROBABLY exists...');
  } else {
    print('"{"first-name": "Dart", "last-name": "lang"}" DEFINITELY does not exist...');
  }
  if (bloom.check('{"first-name": "Go", "last-name": "lang"}') == true) {
    print('"{"first-name": "Go", "last-name": "lang"}" PROBABLY exists...');
  } else {
    print('"{"first-name": "Go", "last-name": "lang"}" DEFINITELY does not exist...');
  }
}


example/bloom_filter_example.dart
import 'package:bloom_filter/bloom_filter.dart';

void main() {
  var bloom = BloomFilter(10000, 3);

  bloom.add('hello world');
  bloom.add(1234567890.toString());
  bloom.add('{"first-name": "Dart", "last-name": "lang"}');

  if (bloom.check('hello world') == true) {
    print('"hello world" PROBABLY exists...');
  }
  else {
    print('"hello world" DEFINITELY does not exist...');
  }
  if (bloom.check(1234567890.toString()) == true) {
    print('"1234567890" PROBABLY exists...');
  }
  else {
    print('"1234567890" DEFINITELY does not exist...');
  }
  if (bloom.check('{"first-name": "Dart", "last-name": "lang"}') == true) {
    print('"{"first-name": "Dart", "last-name": "lang"}" PROBABLY exists...');
  }
  else {
    print('"{"first-name": "Dart", "last-name": "lang"}" DEFINITELY does not exist...');
  }
  if (bloom.check('{"first-name": "Go", "last-name": "lang"}') == true) {
    print('"{"first-name": "Go", "last-name": "lang"}" PROBABLY exists...');
  }
  else {
    print('"{"first-name": "Go", "last-name": "lang"}" DEFINITELY does not exist...');
  }
}
